src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs

```diff
@@ -118,20 +107,40 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
             }
         }
 
-        private async Task<AuthenticationResult> FetchNewAccessTokenAsync(CancellationToken cancellationToken)
+        private async Task<AuthenticationResult> FetchNewAccessTokenAsync(CancellationToken cancellationToken, bool proactivelyRefreshed = false)
         {
             await ResolveAuthorityAsync().ConfigureAwait(false);
             MsalTokenResponse msalTokenResponse;
-            if (ServiceBundle.Config.AppTokenProvider != null)
-            {
-                msalTokenResponse = await SendTokenRequestToProviderAsync(cancellationToken).ConfigureAwait(false);
-            }
-            else
+
+            if (ServiceBundle.Config.AppTokenProvider == null)
             {
                 msalTokenResponse = await SendTokenRequestAsync(GetBodyParameters(), cancellationToken).ConfigureAwait(false);
+                return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
             }
 
-            return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
+            //allow only one call to the provider 
+            await s_semaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                MsalAccessTokenCacheItem cachedAccessTokenItem = await TryGetCachedAccessTokenAsync().ConfigureAwait(false);
+
+                if (cachedAccessTokenItem != null && !_clientParameters.ForceRefresh && 
+                    !proactivelyRefreshed && string.IsNullOrEmpty(AuthenticationRequestParameters.Claims))
+                {
+                    var authResult = CreateAuthenticationResultFromCache(cachedAccessTokenItem);
+                    return authResult;
+                }
+                else
+                {
+                    msalTokenResponse = await SendTokenRequestToProviderAsync(cancellationToken).ConfigureAwait(false);
+                    return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                s_semaphoreSlim.Release();
```

Comment:
Are you adding this semaphore to vanilla `client_credentials` as well? I don't think this is right, as it can dramatically change the behavior of existing apps.

Reply:
not being added to the `client_credentials`, this gets executed only when `.WithAuthority("https://login.microsoftonline.com/tid")` is used on the cca flow